pkg/server/service/loadbalancer/wrr/wrr.go

```diff
@@ -37,79 +42,69 @@ func New(sticky *dynamic.Sticky) *Balancer {
 	return balancer
 }
 
-// Balancer is a WeightedRoundRobin load balancer.
-type Balancer struct {
-	handlers []*namedHandler
-	mutex    *sync.Mutex
-	// Current index (starts from -1)
-	index         int
-	currentWeight int
-	stickyCookie  *stickyCookie
+type namedHandlers []*namedHandler
+
+// Len implements heap.Interface/sort.Interface
+func (n namedHandlers) Len() int { return len(n) }
+
+// Less implements heap.Interface/sort.Interface
+func (n namedHandlers) Less(i, j int) bool {
+	return n[i].deadline < n[j].deadline
 }
 
-func (b *Balancer) maxWeight() int {
-	max := -1
-	for _, s := range b.handlers {
-		if s.weight > max {
-			max = s.weight
-		}
-	}
-	return max
+// Swap implements heap.Interface/sort.Interface
+func (n namedHandlers) Swap(i, j int) {
+	n[i], n[j] = n[j], n[i]
 }
 
-func (b *Balancer) weightGcd() int {
-	divisor := -1
-	for _, s := range b.handlers {
-		if divisor == -1 {
-			divisor = s.weight
-		} else {
-			divisor = gcd(divisor, s.weight)
-		}
+// Push implements heap.Interface for pushing an item into the heap
+func (n *namedHandlers) Push(x interface{}) {
+	h, ok := x.(*namedHandler)
+	if !ok {
+		return
 	}
-	return divisor
+	*n = append(*n, h)
 }
 
-func gcd(a, b int) int {
-	for b != 0 {
-		a, b = b, a%b
-	}
-	return a
+// Pop implements heap.Interface for poping an item from the heap
+func (n *namedHandlers) Pop() interface{} {
+	old := *n
+	h := old[len(old)-1]
+	*n = old[0 : len(old)-1]
+	return h
+}
+
+// Balancer is a WeightedRoundRobin load balancer based on Earliest Deadline First (EDF).
+// (https://en.wikipedia.org/wiki/Earliest_deadline_first_scheduling)
+// Each pick from the schedule has the earliest deadline entry selected. Entries have deadlines set
+// at currentDeadline + 1 / weight, providing weighted round robin behavior with floating point
+// weights and an O(log n) pick time.
+type Balancer struct {
+	handlers     *namedHandlers
+	mutex        *sync.Mutex
+	curDeadline  float64
+	curLock      sync.RWMutex
```

Comment:
why did you add a new mutex?
I think you can simply change mutex L84 to a RWMutex, and use it to guard all the fields in Balancer that require it, no? and then you can remove curLock.